pkg/controller/azvolumeattachment.go

```diff
@@ -0,0 +1,697 @@
+	*/
+	// syncMutex is used to prevent other syncVolume calls to be performed during SyncAll routine
+	syncMutex sync.RWMutex
+	// muteMap maps volume name to mutex, it is used to guarantee that only one sync call is made at a time per volume
+	mutexMap map[string]*sync.Mutex
+	// muteMapMutex is used when updating or reading the mutexMap
+	mutexMapMutex sync.RWMutex
+	// volueMap maps AzVolumeAttachment name to volume, it is only used when an AzVolumeAttachment is deleted,
+	// so that the controller knows which volume to manage replica for in next iteration of reconciliation
+	volumeMap map[string]string
+	// volumeMapMutex is used when updating oreading the volumeMap
+	volumeMapMutex sync.RWMutex
+}
+
+type filteredNode struct {
+	azDriverNode v1alpha1.AzDriverNode
+	numAttached  int
+}
+
+var _ reconcile.Reconciler = &reconcileAzVolumeAttachment{}
+
+func (r *reconcileAzVolumeAttachment) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	return r.HandleAzVolumeAttachmentEvent(ctx, request)
+}
+
+func (r *reconcileAzVolumeAttachment) HandleAzVolumeAttachmentEvent(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	var azVolumeAttachment v1alpha1.AzVolumeAttachment
+	err := r.client.Get(ctx, request.NamespacedName, &azVolumeAttachment)
+	// if object is not found, it means the object has been deleted. Log the deletion and do not requeue
+	if errors.IsNotFound(err) {
+		klog.Infof("AzVolumeAttachment (%s) has successfully been deleted.", request.Name)
+		r.volumeMapMutex.RLock()
+		underlyingVolume, ok := r.volumeMap[request.Name]
+		r.volumeMapMutex.RUnlock()
+		if ok {
+			if err = r.ManageReplicas(ctx, underlyingVolume); err != nil {
+				klog.Errorf("failed to manage replicas for volume (%s): %v", underlyingVolume, err)
+				return reconcile.Result{Requeue: true}, err
+			}
+			// delete (azVolumeAttachment, volume) from the volumeMap
+			r.volumeMapMutex.Lock()
+			delete(r.volumeMap, request.Name)
+			r.volumeMapMutex.Unlock()
+		}
+		return reconcile.Result{}, nil
+		// if the GET failure is not triggered by not found error, log it and requeue the request
+	} else if err != nil {
+		klog.Errorf("failed to fetch azvolumeattachment object with namespaced name %s: %v", request.NamespacedName, err)
+		return reconcile.Result{Requeue: true}, err
+	}
+
+	// this is a creation event
+	if azVolumeAttachment.Status == nil {
+		// attach the volume to the specified node
+		klog.Infof("Initiating Attach operation for AzVolumeAttachment (%s)", azVolumeAttachment.Name)
+		if err := r.TriggerAttach(ctx, azVolumeAttachment.Name); err != nil {
+			klog.Errorf("failed to attach AzVolumeAttachment (%s): %v", azVolumeAttachment.Name, err)
+			return reconcile.Result{Requeue: true}, err
+		}
+		// if the azVolumeAttachment's deletion timestamp has been set, detach the disk from the node and delete the finalizer
+	} else if !azVolumeAttachment.ObjectMeta.DeletionTimestamp.IsZero() {
```

Comment:
We should check if the `DeletionTimestamp` is >= the current time and not just zero. Kubernetes allows for deletion to be scheduled at a later time.

Reply:
Good point. I have updated the clause such that it checks whether `DeletionTimestamp` is `before` the current time.